### DISCUSSION TITLE
2022-11: 新轮子 reqsign

### DISCUSSION THREAD

**Discussion by giscus** [Xuanwo/blog#343] [Announcements]
# 2022-11: 新轮子 reqsign

An infrastructure engineer, focused on distributed storage system

https://xuanwo.io/reports/2022-11/

**Comment by xxchan** [comment]
怎么感觉有点无敌啊🤔那做跨☁应用是不是再也不用SDK了

**Reply by Xuanwo** [reply]
如果都是跟 S3 这样简单的 API 问题不大，像 github 那种复杂还是太麻烦了

**Comment by laixintao** [comment]
分享一个今年才知道的一个很神奇的设计。这个问题的根本是，1）希望 API 的 SDK 可以让专业的人去维护，不要用户维护 2）希望减少依赖，安装的时候不要装太多的东西。

有一种解决方案是，（基本在公司里才会这么用），公司的一个团队提供 agent，会运行在每一台机器上，这个 agent 接受 rpc 协议的请求，然后在后端对接不同的系统，比如 MySQL，Redis，S3，其他 Service，外部 Service 等等……

用户原先需要安装这些软件对应的 SDK (mysqlclient, redis client, s3 client...) 发送请求至这些服务，现在变成了用户只需要安装 agent 的 rpc sdk，然后将请求发送到 agent 上，client app (rpc)->agent -> MySQL.

这样似乎可以解决本文提出的问题。

好处是：用户只需要安装一个依赖，就可以访问公司里面的大部分服务；依赖的管理最终回到了一个专业团队维护，这个专业团队不会面对”用户侧的 SDK 难以升级“的问题，因为他们可以统一维护 agent 的升级。

坏处也很明显：用户无法使用没有被公司统一支持的依赖；并且，比如用户需要用一个基于 redis 的 job queue, 但是这个 job queue 是使用 redis sdk 来访问 redis 的，那么必须有人 fork 修改，将它变成 rpc 请求，才能使用。不过对于一个公司里面成百上千的服务来说，这点成本应该是可以接受的。而且还可以控制用户使用的依赖（服务治理）。

**Reply by Xuanwo** [reply]
有点像是 service mesh 想搞定的事情

**Comment by wey-gu** [comment]
> 假设我们有一个库，能把上面的这些跟请求签名相关的工作全部搞定

👍🏻，觉得🌀哥哥这篇文章值得被用英文写一下，然后在 https://reqsign.xuanwo.io/reqsign/ 引用一下，然后再去 hackernews 分享一下
